### PR TITLE
Remove the `periodic` argument from `xgcm.Grid`

### LINKED ISSUE #195: boundary="periodic"?
I keep finding it funny that xgcm has both `periodic` and `boundary` kwargs. Logically it seems like `periodic` should be replaced by `boundary="periodic"`?

### LINKED ISSUE #509: [BUG] Diff in non periodic axis seems to be periodic
**What happened**:

When applying a diff operation on a simple example (modified from the doc https://xgcm.readthedocs.io/en/latest/grids.html), it seems that the X axis is considered as periodic even if stated otherwise.

**What you expected to happen**:

In our minimal example, we apply the diff operation on a small array [2, 4, 5, 8, 10, 12, 14, 16, 18]. We expected to have [np.nan, 2, 2, 2...] or an array with one less element, but we got [-16, 2, 2, 2...] instead. It seems that the last element is taken to extend the boundary at the left, whereas it has no relation with the first element whatsoever.

**Minimal Complete Verifiable Example**:

```python
import xarray as xr
import xgcm
import numpy as np

ds = xr.Dataset(
	coords={
		"x_c": (
			["x_c"],
			np.arange(1, 10),
		),
		"x_g": (
			["x_g"],
			np.arange(0.5, 9),
		),
	}
)

da = ds.x_c*2
grid = xgcm.Grid(ds, coords={"X": {"center": "x_c", "left": "x_g"}}, periodic=False)

grid
#>> <xgcm.Grid>
#>> X Axis (not periodic, boundary=None):
#>>  * center   x_c --> left
#>>  * left     x_g --> center

grid.diff(da, "X")
#>> <xarray.DataArray 'x_c' (x_g: 9)>
#>> array([-16,   2,   2,   2,   2,   2,   2,   2,   2])
#>> Coordinates:
#>>   * x_g      (x_g) float64 0.5 1.5 2.5 3.5 4.5 5.5 6.5 7.5 8.5
```

**xgcm version and environment**:

xgcm 0.7.0

<details>

</details>

### LINKED ISSUE #604: [BUG] Default boundary condition is periodic even when specified otherwise
The issue: Even when xgcm has been told that a coordinate is not periodic, it still chooses that as the default behavior - instead of throwing an error or warning. 

```
import xarray as xr 
import xgcm
import numpy as np

dx = 0.5
L = 20
x_left = np.arange(0,L-dx, dx)
x = np.arange(dx, L, dx)

slope = 2.
y = slope*x

da_y = xr.DataArray(y, dims=['xc'], coords={'xc':x})
ds_y = xr.Dataset(coords={'xc':x, 'xg':x_left})
ds_y['y'] = da_y

grid = xgcm.Grid(ds_y, 
                 coords={"X": {"center": "xc", "left": "xg"}},
                periodic=None)

dy_default = grid.diff(ds_y.y, 'X')
dy = grid.diff(ds_y.y, 'X', boundary='extend')
```

In this example we told grid that `X` is not a periodic direction. When we print grid, it says that `X` is not periodic. However, the dy_default (where we did not pass a boundary condition) still assumes `X` is periodic without alerting the user.

### LINKED ISSUE #624: [BUG] Non-intuitive behavior if periodic not set as list
If the periodic argument to Grid is not passed as a list then the resultant grid object has non-intuitive result. 

In example below user may expect both object to have only X dim as periodic. However, in case where list is given X dim is periodic and Z is non-periodic, while when the dimension is specified as string the option True is likely being used. 

```
import xarray as xr
import numpy as np
from xgcm import Grid

ds = xr.Dataset(
    coords={
        "zl": (["zl"], np.arange(1,15)),
        "zi": (["zi"], np.arange(0.5, 15.5)),
        "xl": (["xl"], np.arange(1,15)),
        "xi": (["xi"], np.arange(0.5, 15.5)),
    }
)

grid_list = Grid(ds, 
            coords={"Z": {"center": 'zl', 'outer': 'zi'},
                    "X": {"center": 'xl', 'outer': 'xi'}},
           periodic=['X'])

grid_non_list = Grid(ds, 
            coords={"Z": {"center": 'zl', 'outer': 'zi'},
                    "X": {"center": 'xl', 'outer': 'xi'}},
           periodic='X')
```

The solution would be to raise an error to the user and force them to use a list, or automatically convert string to a list of a single value.

### LINKED ISSUE #625: [BUG] cumsum seems to apply periodic B.C. even when it knows that is not the case.
We should expect that for non-periodic axis cumsum should start integration with first value as 0 (if using outer or left for the axis being cumsumed to), and increase or decrease accordingly.

Instead, the current implementation seems to be applying periodic conditions and filling the first value with the sum. 

```
ds = xr.Dataset(
    coords={
        "zl": (["zl"], np.arange(1,15)),
        "zi": (["zi"], np.arange(0.5, 15.5))
    }
)

grid = Grid(ds, 
            coords={"Z": {"center": 'zl', 'outer': 'zi'}},
           periodic=False)

temp = grid.cumsum(ds.zl, 'Z')
temp1 = grid.cumsum(ds.zl, 'Z', boundary='fill', fill_value=0.)
```

Both cumsums start from 0 and increase, instead the output is: 
Screen Shot 2024-03-27 at 2 48 31 PM

As seen above, when explicit B.C. is not provided we end up with very large value at start.

### PR DESCRIPTION
**Pull request by hdrake**
## What

Removes the `periodic` argument of `xgcm.Grid`. Boundary behavior is now controlled exclusively by `boundary`. Passing `periodic=` raises an informative `ValueError` pointing at the replacement (per the deprecation policy from #696).

## Why

`periodic` defaulted to `True`, silently making every axis periodic. That default was the root of several boundary-condition bugs (#509, #604, #624) and created confusion between the `periodic` string/list forms and `boundary` (#624). Removing it makes boundary handling explicit and non-periodic by default. Part of the v1.0.0 push (#733, roadmap §A/§B); scope was discussed there and in #734, and the related `boundary`→`padding` rename is kept separate in #696 to keep this diff reviewable.

## New default semantics (maintainer-review point)

An axis with no `boundary` specified now applies **no boundary condition** (`boundary=None`, which is what the docstring always claimed). Operations that need to pad such an axis raise an informative error instead of silently wrapping; an op-level `boundary=` still works as before. Pass `boundary='periodic'` at grid construction to recover the pre-1.0 behavior. This also fixes cases where a declared non-periodic axis could still wrap.

**Face-connection carve-out** (from review): a **fully connected** edge does not require a boundary — for grids like a global cubed sphere, where every padded edge's halo is overwritten by connection data, operations work without any `boundary` (matching master's numerics). The informative error still fires when a *genuinely unconnected* edge of a partially-connected grid has no boundary at either the grid or op level.

## Migration

- `periodic=True` → `boundary='periodic'`
- `periodic=False` → `boundary='fill'` (the previous implicit mapping)
- `periodic=['X']` → `boundary={'X': 'periodic', 'Y': 'fill'}` (per-axis dict)
- Fully face-connected grids need no boundary; otherwise rely on the default only if every operation either doesn't pad or is called with an explicit `boundary=`.

## Tests

Migrated the whole suite and the executed docs off `periodic=`, preserving each test's intent (grids that relied on the old periodic default got explicit `boundary="periodic"` to preserve exact numerics). Added regression tests for the removal `ValueError` and for #509/#604/#624; `test_diff_interp_cubed_sphere` again builds its grid with **no** boundary (as on master) plus a hand-derived connected-halo assertion, and `test_unconnected_edge_without_boundary_raises` covers the partially-connected error path. Also fixes a latent test bug where `test_metrics_ops` parametrized boundary as the strings `"True"`/`"False"` (both silently mapped to `fill`). The migration also covers `test_conservative_transform_explicit_target_dim`, added on master after this branch was cut (it still passed `periodic=False`, which broke CI post-merge).

As part of the AI iteration described under *How AI was used*, adversarial review passes were used to check that the test-collection delta vs master is fully accounted for (coverage improved, none lost), that the op-level `boundary=` override still works at runtime, and that the #509/#604 regression test fails on master. These informed — but do not replace — the line-by-line human review I'm completing separately (to be summarized in a comment below).

Full suite green (`pixi run tests`), `pixi run lint` clean, `pixi run -e docs docs-build` clean.

Closes #195, closes #509, closes #604, closes #624, closes #625. Supersedes #626 — earlier work by @jbusecke.

## How AI was used

I used [Claude Code](https://claude.com/claude-code) (Fable 5 model) to draft this PR. I pointed it towards @jbusecke 's earlier PR #626 and asked it to pick up that work but merging in the current upstream/master. Since this PR is conceptually pretty straight-forward but requires a large diff, I think it is a good candidate for off-loading to AI. It helps that there are multiple issues summarizing the motivation for this PR and an entirely human-written PR #626 that started the work.

I iterated with it a little bit, e.g. by asking it to have independent review agents review it to make sure no instances of `periodic` were missed. Most of these changes were done in a single commit, with only minor iterations later to add guards and tests. I reviewed the summaries of code changes as Claude was making them, but will conduct a full line-by-line review separately later (and summarize that in a comment on this PR).

### AI Disclosure

- [x] This PR contains AI-generated content.
  - [x] I have tested any AI-generated content in my PR.
  - [x] I take responsibility for any AI-generated content in my PR.
    Tools: Claude Code (Claude Fable 5)

### CODE REVIEW

File: xgcm/test/test_grid.py [hunk 1]
```diff
@@ -384,7 +452,9 @@ def test_keep_coords_removed():
 
     ds, coords, metrics = datasets_grid_metric("B")
     ds = ds.assign_coords(yt_bis=ds["yt"], xt_bis=ds["xt"])
-    grid = Grid(ds, coords=coords, metrics=metrics, autoparse_metadata=False)
+    grid = Grid(
+        ds, coords=coords, metrics=metrics, boundary="fill", autoparse_metadata=False
+    )
```
Comment (hdrake) [inline xgcm/test/test_grid.py:630]: Why does this get `boundary="fill"` while R422-428 get `boundary="periodic"` if both are meant to replicate the original default behavior?

Comment (hdrake) [reply xgcm/test/test_grid.py:630]: Good catch — inconsistent, but harmless. On master both tests relied on the bare default (`periodic=True`). `test_keep_coords` was migrated to `boundary="periodic"` because it actually runs padding ops whose numerics depend on it. `test_keep_coords_removed` never pads, though — the `keep_coords` guard raises at the top of `diff`/`cumsum`/`apply_as_grid_ufunc`, before any boundary handling — so `"fill"` here was arbitrary. Switched it to `boundary="periodic"` to match the sibling and the default→periodic migration rule.

🤖 _AI-generated with [Claude Code](https://claude.com/claude-code); reviewed by @hdrake._

**Comment by hdrake** [comment]
I reviewed all of the line-by-line changes as well as the updated docs. This looks good to me.
